### PR TITLE
Added post-quantum client/server pubkey auth.

### PR DESCRIPTION
**Pull request by christianpaquin**
Initial commit of post-quantum authentication. This adds pubkey generation and authentication to both the client and server, using the master-branch liboqs algorithms.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -206,7 +245,7 @@ The purpose of this option is as follows.  liboqs master branch and liboqs nist-
 	- `make clean`
 	- `make -j`
 	- `make install`
-3. Run `ssh` with `ecdh-nistp384-oqsdefault-sha384@openquantumsafe.org ` or `oqsdefault-sha384@openquantumsafe.org` for the `KexAlgorithms` option.
+3. Run `ssh` with `ecdh-nistp384-oqsdefault-sha384@openquantumsafe.org ` or `oqsdefault-sha384@openquantumsafe.org` for the `KexAlgorithms` option, or +++
```
Comment (dstebila) [inline README.md]: "or +++" -- presumably more text to add here?

**Review by dstebila — COMMENTED** [review]
liboqs nist branch uses a different signature API.  Does this work only against liboqs master branch then?  If so we should be clear about that around line 98 of README.md.

Also, we should update the integration test script oqs\_openssh.sh to exercise the new code.

File: README.md (hunk 1 above)
Comment (christianpaquin) [reply README.md]: Yes, missing text; I'll add it.

**Comment by himsen** [comment]
Hi,

A general comment: it is possible to disable individual algorithms in liboqs. So, you should verify in `configure.ac` that all algorithms you want to offer are not disabled.

**Comment by christianpaquin** [comment]
@himsen; I saw the KEX code with per-alg ifdefs, but in this initial commit, I decided to follow the OpenSSL integration strategy, and have a single WITH_OQS macro instead of a per-scheme one. I feel the per-scheme one might be a bit heavy, and means that the OpenSSH code would need to be updated often, every time OQS changes. A disabled OQS algorithm can be handled by OpenSSH reporting properly the null return value from the init calls.

**Comment by christianpaquin** [comment]
@dstebila, I updated the README. As discussed in the dev meeting, it would be good if you or someone else could update the integration test to test my documented steps.

**Comment by dstebila** [comment]
I got this working locally on my Mac with liboqs master branch.  I've started working on the integration tests, but then realized that this doesn't compile against liboqs nist-branch because of the different signature API.  

How should we proceed?  

- Use `#ifdef`s to disable PQ auth when building against liboqs nist-branch?
- Forget about building OpenSSH against liboqs nist-branch?

**Comment by christianpaquin** [comment]
> How should we proceed?
> 
> * Use #ifdefs to disable PQ auth when building against liboqs nist-branch?
> * Forget about building OpenSSH against liboqs nist-branch?

The former option seems better. I'll change the WITH_OQS macro wrapping the auth code to WITH_OQS_AUTH, and add a configure option to enable it (like for KEX).

**Comment by christianpaquin** [comment]
@dstebila; I've push a commit adding the enable-pq-auth option, addressing your former comment.

**Review by dstebila — APPROVED** [review]

**Comment by dstebila** [comment]
Passes my local integration tests.  Will be added to integration testing framework via open-quantum-safe/testing#27.

**Pull request merged**